### PR TITLE
Add ClrArray.GetStructValue()

### PR DESCRIPTION
**Pull request by chrisnas**
Add GetStructValue() to ClrArray and refactor GetObjectValue() accordingly.
Should fix https://github.com/microsoft/clrmd/issues/786

### CODE REVIEW

File: src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs [hunk 1]
```diff
@@ -157,15 +157,65 @@ public int GetUpperBound(int dimension)
...
+            }
+            else
+            {
+                return Address + (ulong)valueByteOffset;
+            }
+        }
+
+        public ClrValueType GetStructValue(int index)
+        {
+            if (Type.ComponentType.IsObjectReference)
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs:201]: ComponentType needs to be null checked.  Unfortunately there's some types we just can't get information on, and an array of generic types sometimes has a null ComponentType.  I think returning an empty ClrValueType (which will has `ClrValueType.IsValid == false`) is the only sensible thing to do in that case.

File: src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs [hunk 2]
```diff
@@ -157,15 +157,65 @@ public int GetUpperBound(int dimension)
...
+
+            ulong address = GetItemAddress(index, false);
+            return new ClrValueType(address, Type.ComponentType, false);
+        }
+
+
         public ClrObject GetObjectValue(int index)
         {
-            if (!Type.IsObjectReference)
+            if (!Type.ComponentType.IsObjectReference)
```
Comment (leculver) [inline src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs:211]: This needs a null check, build warning.  Same as above.

File: src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs [hunk 3]
```diff
@@ -157,15 +157,65 @@ public int GetUpperBound(int dimension)
...
+            {
+                return Type.ClrObjectHelpers.DataReader.Read<UIntPtr>(Address + (ulong)valueByteOffset).ToUInt64();
+            }
+            else
+            {
+                return Address + (ulong)valueByteOffset;
+            }
+        }
+
+        public ClrValueType GetStructValue(int index)
```
Comment (NextTurn) [inline src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs:199]: `struct` is merely a C# concept.

File: src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs [hunk 4]
```diff
@@ -157,15 +157,65 @@ public int GetUpperBound(int dimension)
 
         public T GetValue<T>(params int[] indices) where T : unmanaged => GetValue<T>(false, indices);
 
+        // This helper method returns the address of the item at a given index of an array.
+        // - reference type object: the array item at the given index contains the address so it has to be read.
+        // - value type object: it will be embedded inside the array itself so the address is computed relatively to the array itself.
+        private ulong GetItemAddress(int index, bool isObjectReference)
```
[the remark this answers is not in the document]
Comment (NextTurn) [reply src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs:163]: This has been covered by `(uint)valueOffset` checks below.

File: src/Microsoft.Diagnostics.Runtime.Tests/src/BasicArrayTests.cs [hunk 5]
```diff
@@ -27,6 +27,7 @@ public BasicArrayTests(ArrayConnection connection)
             _connection = connection;
         }
 
+
```
Comment (sungam3r) [inline src/Microsoft.Diagnostics.Runtime.Tests/src/BasicArrayTests.cs:30]:
```suggestion
```

File: src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (chrisnas) [reply src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs:163]: I prefer an ArgumentOutOfRangeException to a System.OverflowException : Arithmetic operation resulted in an overflow so I'm fixing it.

[the remark this answers is not in the document]
Comment (NextTurn) [reply src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs:163]: The casts will not be checked.

File: src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs (hunk 1 above)
Comment (chrisnas) [reply src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs:201]: Good idea.

File: src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs (hunk 2 above)
Comment (chrisnas) [reply src/Microsoft.Diagnostics.Runtime/src/Common/ClrArray.cs:211]: ok

File: src/Microsoft.Diagnostics.Runtime.Tests/src/BasicArrayTests.cs (hunk 5 above)
Comment (chrisnas) [reply src/Microsoft.Diagnostics.Runtime.Tests/src/BasicArrayTests.cs:30]: line removed

**Comment by leculver** [comment]
LGTM.

**Pull request merged**